File: autosew/autosew/curves.py
"""Edge polyline reconstruction and the sagitta shape descriptor.

The tagged-union curvature encoding (k_t + k1..k10, where k_t switches the meaning of
the ten parameter slots) is fine when k_t comes from a specification file and is exact.
It is not usable when the input is a DXF polyline, where k_t has to be *estimated*: a
circular arc and a weakly-curved quadratic Bezier fit sampled points about equally well,
so k_t flips, and eleven feature dimensions change meaning at once.

The sagitta profile removes the type decision entirely.  Both a specification curve and a
DXF polyline are reduced to the same thing -- signed perpendicular deviation from the
chord, at K uniform arclength positions, divided by the chord length:

    v_k = ((p(s_k) - p0) . n) / L ,   s_k = k/(K+1) of the arc,  k = 1..K

with n the left normal of the chord in the panel's ACW traversal frame.  The values are
dimensionless, always mean the same thing, and vary continuously with the shape, so a
refitting error is a small perturbation rather than a categorical error.  Two edges that
get stitched together have equal length and mirror-image curvature, which in this
encoding is a sign flip plus a reversal -- directly visible to an inner product.
"""
from __future__ import annotations
import math

import numpy as np

from .config import KT_STRAIGHT, KT_CIRCLE, KT_QUADRATIC, KT_CUBIC, KT_BSPLINE

SAMPLES_PER_EDGE = 32   # curve -> polyline resolution before resampling


def arc_polyline(p0, p1, r, large, sweep, n=SAMPLES_PER_EDGE):
    """SVG-style circular arc, same convention as viz/visualize.py:arc_points.
    GCD circle params are [radius, large_arc, right]."""
    chord = p1 - p0
    d = float(np.linalg.norm(chord))
    if d < 1e-9:
        return np.stack([p0, p1])
    r = max(float(r), d / 2 + 1e-9)
    mid = (p0 + p1) / 2
    h = math.sqrt(max(r * r - d * d / 4, 0.0))
    perp = np.array([-chord[1], chord[0]]) / d
    center = mid + (h if bool(large) != bool(sweep) else -h) * perp
    a0 = math.atan2(p0[1] - center[1], p0[0] - center[0])
    a1 = math.atan2(p1[1] - center[1], p1[0] - center[0])
    if sweep:
        while a1 < a0:
            a1 += 2 * math.pi
    else:
        while a1 > a0:
            a1 -= 2 * math.pi
    t = np.linspace(a0, a1, n)
    return np.stack([center[0] + r * np.cos(t), center[1] + r * np.sin(t)], axis=1)


def edge_polyline(edge, n=SAMPLES_PER_EDGE):
    """Parsed Edge -> polyline in the same panel-local frame as edge.start/edge.end.

    Bezier control points in edge.kparams are already absolute and already translated
    with the panel bbox, so they need no further conversion.  B-splines fall back to the
    chord (the parameter layout is not pinned down; they do not occur in GCD.v2 specs
    seen so far -- callers that care should count edge.kt == KT_BSPLINE)."""
    p0 = np.asarray(edge.start, float)
    p1 = np.asarray(edge.end, float)
    kp = edge.kparams
    if edge.kt == KT_QUADRATIC and len(kp) >= 2:
        q = np.array(kp[:2], float)
        t = np.linspace(0, 1, n)[:, None]
        return (1 - t) ** 2 * p0 + 2 * (1 - t) * t * q + t ** 2 * p1
    if edge.kt == KT_CUBIC and len(kp) >= 4:
        q1 = np.array(kp[0:2], float)
        q2 = np.array(kp[2:4], float)
        t = np.linspace(0, 1, n)[:, None]
        return ((1 - t) ** 3 * p0 + 3 * (1 - t) ** 2 * t * q1
                + 3 * (1 - t) * t ** 2 * q2 + t ** 3 * p1)
    if edge.kt == KT_CIRCLE and len(kp) >= 1:
        large = kp[1] if len(kp) > 1 else 0
        sweep = kp[2] if len(kp) > 2 else 0
        return arc_polyline(p0, p1, kp[0], large, sweep, n)
    return np.stack([p0, p1])


def sagitta_profile(pts, K):
    """Polyline -> (K,) signed perpendicular deviation from the chord, divided by the
    chord length, sampled at uniform arclength fractions k/(K+1).

    Endpoints are excluded: their deviation is identically zero and carries nothing."""
    pts = np.asarray(pts, float)
    out = np.zeros(K, dtype=np.float64)
    if len(pts) < 3:
        return out
    p0, p1 = pts[0], pts[-1]
    chord = p1 - p0
    L = float(np.hypot(chord[0], chord[1]))
    if L < 1e-12:
        return out
    nrm = np.array([-chord[1], chord[0]]) / L      # left normal, ACW frame
    seg = np.linalg.norm(np.diff(pts, axis=0), axis=1)
    s = np.concatenate([[0.0], np.cumsum(seg)])
    if s[-1] < 1e-12:
        return out
    s /= s[-1]
    targets = np.arange(1, K + 1) / (K + 1.0)
    qx = np.interp(targets, s, pts[:, 0])
    qy = np.interp(targets, s, pts[:, 1])
    return ((np.stack([qx, qy], 1) - p0) @ nrm) / L

File: webdemo/src/lib/curves.ts
/* Port of autosew/autosew/curves.py -- edge polyline reconstruction and the sagitta
   shape descriptor.  Keep the two in step: the model was trained on exactly these
   numbers, so any divergence silently degrades the prediction rather than raising.

   The tagged curvature encoding (k_t plus ten slots whose meaning k_t selects) is fine
   when k_t comes from a specification file and is exact.  It is not usable when the
   input is a DXF polyline, where k_t has to be estimated: a circular arc and a weakly
   curved quadratic Bezier fit sampled points about equally well, so k_t flips and
   eleven feature dimensions change meaning at once.

   The sagitta profile removes the type decision.  Both a specification curve and a DXF
   polyline reduce to the same thing -- signed perpendicular deviation from the chord, at
   K uniform arclength positions, divided by the chord length. */
import { KT, type Edge, type Pt } from "./parseSpec";

export const SAMPLES_PER_EDGE = 32;

/** SVG-style circular arc.  GCD circle params are [radius, large_arc, right]. */
export function arcPolyline(p0: Pt, p1: Pt, r: number, large: number, sweep: number,
                            n = SAMPLES_PER_EDGE): Pt[] {
  const cx = p1[0] - p0[0], cy = p1[1] - p0[1];
  const d = Math.hypot(cx, cy);
  if (d < 1e-9) return [p0, p1];
  const rr = Math.max(r, d / 2 + 1e-9);
  const h = Math.sqrt(Math.max(rr * rr - (d * d) / 4, 0));
  const px = -cy / d, py = cx / d;
  const s = (!!large) !== (!!sweep) ? 1 : -1;
  const c: Pt = [(p0[0] + p1[0]) / 2 + s * h * px, (p0[1] + p1[1]) / 2 + s * h * py];
  let a0 = Math.atan2(p0[1] - c[1], p0[0] - c[0]);
  let a1 = Math.atan2(p1[1] - c[1], p1[0] - c[0]);
  if (sweep) { while (a1 < a0) a1 += 2 * Math.PI; } else { while (a1 > a0) a1 -= 2 * Math.PI; }
  return Array.from({ length: n }, (_, k) => {
    const t = a0 + ((a1 - a0) * k) / (n - 1);
    return [c[0] + rr * Math.cos(t), c[1] + rr * Math.sin(t)] as Pt;
  });
}

/** Parsed Edge -> polyline in the same panel-local frame as edge.start / edge.end.
    Bezier control points in kparams are already absolute and already translated with
    the panel bbox, so they need no further conversion.  B-splines fall back to the
    chord (their parameter layout is not pinned down and they do not occur in the
    GCD.v2 specs seen so far). */
export function edgePolyline(e: Edge, n = SAMPLES_PER_EDGE): Pt[] {
  if (e.poly) return e.poly;      // DXF input: the sampled boundary IS the curve
  const p0 = e.start, p1 = e.end, kp = e.kparams;
  if (e.kt === KT.QUADRATIC && kp.length >= 2) {
    const q: Pt = [kp[0], kp[1]];
    return Array.from({ length: n }, (_, k) => {
      const t = k / (n - 1), u = 1 - t;
      return [u * u * p0[0] + 2 * u * t * q[0] + t * t * p1[0],
              u * u * p0[1] + 2 * u * t * q[1] + t * t * p1[1]] as Pt;
    });
  }
  if (e.kt === KT.CUBIC && kp.length >= 4) {
    const q1: Pt = [kp[0], kp[1]], q2: Pt = [kp[2], kp[3]];
    return Array.from({ length: n }, (_, k) => {
      const t = k / (n - 1), u = 1 - t;
      return [u * u * u * p0[0] + 3 * u * u * t * q1[0] + 3 * u * t * t * q2[0] + t * t * t * p1[0],
              u * u * u * p0[1] + 3 * u * u * t * q1[1] + 3 * u * t * t * q2[1] + t * t * t * p1[1]] as Pt;
    });
  }
  if (e.kt === KT.CIRCLE && kp.length >= 1) {
    return arcPolyline(p0, p1, kp[0], kp.length > 1 ? kp[1] : 0, kp.length > 2 ? kp[2] : 0, n);
  }
  return [p0, p1];
}

/** Polyline -> K signed perpendicular deviations from the chord, over the chord length,
    sampled at uniform arclength fractions k/(K+1).  Endpoints are excluded: their
    deviation is identically zero and carries nothing. */
export function sagittaProfile(pts: Pt[], K: number, out?: Float32Array, off = 0): Float32Array {
  const o = out ?? new Float32Array(K);
  if (!out) o.fill(0);
  else for (let k = 0; k < K; k++) o[off + k] = 0;
  if (pts.length < 3) return o;

  const p0 = pts[0], p1 = pts[pts.length - 1];
  const cx = p1[0] - p0[0], cy = p1[1] - p0[1];
  const L = Math.hypot(cx, cy);
  if (L < 1e-12) return o;
  const nx = -cy / L, ny = cx / L;                 // left normal, ACW frame

  const s = new Float64Array(pts.length);
  for (let i = 1; i < pts.length; i++)
    s[i] = s[i - 1] + Math.hypot(pts[i][0] - pts[i - 1][0], pts[i][1] - pts[i - 1][1]);
  const total = s[pts.length - 1];
  if (total < 1e-12) return o;

  let j = 1;
  for (let k = 0; k < K; k++) {
    const t = ((k + 1) / (K + 1)) * total;
    while (j < pts.length - 1 && s[j] < t) j++;
    const span = s[j] - s[j - 1];
    const w = span > 1e-15 ? (t - s[j - 1]) / span : 0;
    const qx = pts[j - 1][0] + w * (pts[j][0] - pts[j - 1][0]);
    const qy = pts[j - 1][1] + w * (pts[j][1] - pts[j - 1][1]);
    o[off + k] = ((qx - p0[0]) * nx + (qy - p0[1]) * ny) / L;
  }
  return o;
}
